### PR TITLE
Merge Ray Train XGboost tutorials

### PR DESCRIPTION
**Pull request by crypdick**
## Why are these changes needed?

This PR merges the two existing Ray Train XGBoost tutorials, which were largely duplicated, into one.

* created new distributed-xgboost-lightgbm.ipynb which is a superset of the two existing tutorials
* ported .rst with transcluded code snippets since it was a tangled nightmare
* updated pointers in train/examples.yml to this tutorial
* kept orphaned code in doc_code/preprocessors.py since it is part of our CI

**NOTE**: I am not sure that I updated the pointers to this tutorial correctly. Please double check my work in `examples.yml`

Out of scope: merging the third Ray Tune XGBoost tutorial.

## Related issue number

https://github.com/ray-project/ray/pull/50232#issuecomment-2638339942

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by pcmoritz** [comment]
@crypdick Unfortunately this doesn't work since now none of the links work any more (look at the test failures in https://buildkite.com/ray-project/microcheck/builds/11000) -- these are set by the anchor `.. _train-gbdt-guide:` for example. Either we keep the file in .rst, or we need to figure out how to set the anchor in the .ipynb :)

There was also a small error about a non-existant use case that I pushed a fix for :)

File: doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb
Comment (GokuMohandas) [inline doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: hey @crypdick , can we separate the inference section to accept any dataset as input ([example](https://docs.ray.io/en/latest/data/batch_inference.html#quickstart)). Currently the class `Predict` has a function `predict_xgboost` that explicitly uses the test dataset.

Comment (GokuMohandas) [inline doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: Could we add a section for hyperparameter tuning with these models? We don't need a full blown code example (unless you want to!) but a nice paragraph on how to do it with Ray Tune. I recently found out that it's a very popular workload among our customers (Attentive, slingshot, etc.). 

And one of them (Eric Jeske from Attentive) was asking how to prevent Ray from reloading the data for each new hp tuning trial since his dataset is massive (250tb). We can discuss this last part together because I need to think about it.

Comment (GokuMohandas) [inline doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: Could we add a section for hyperparameter tuning with these models? We don't need a full blown code example (unless you want to!) but a nice paragraph on how to do it with Ray Tune. I recently found out that it's a very popular workload among our users.

And one of them was asking how to prevent Ray from reloading the data for each new hp tuning trial since his dataset is massive (250tb). We can discuss this last part together because I need to think about it.

**Comment by crypdick** [comment]
> @crypdick Unfortunately this doesn't work since now none of the links to the .rst that were converted to .ipynb are functional any more (look at the test failures in https://buildkite.com/ray-project/microcheck/builds/11000) -- these are set by the anchor `.. _train-gbdt-guide:` for example. Either we keep the file in .rst, or we need to figure out how to set the anchor in the .ipynb :)

@pcmoritz I am not sure why the link didn't work. For example, the Ray Data [examples page](https://docs.ray.io/en/latest/data/examples.html) links to an [ipynb tutorial](https://docs.ray.io/en/latest/data/examples/pytorch_resnet_batch_prediction.html) via the [examples.yml links](https://github.com/ray-project/ray/blob/7537aa2620052080d25e13e573d42f04c65c6ef8/doc/source/data/examples.yml#L12) using the same pattern. What am I missing?

File: doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb
Comment (crypdick) [reply doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: to make matters confusing, there is also a 3rd tutorial which is about XGBoost on Ray Tune. I am not sure whether it makes sense to merge that one into the Ray Train tutorial though. What if I created a section for hyperparameter tuning that just links to the existing xgboost+Tune tutorial?

as for the data reloading problem, couldn't this be fixed using Tune's [`reuse_actors` kwarg](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.TuneConfig.html)?

Comment (crypdick) [reply doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: @GokuMohandas can you point me to an example of how we create example tabs within an ipynb file? The example you shared [is an rst file](https://github.com/ray-project/ray/blob/b6e02cc28ae440e1877d7ea130f765da61630040/doc/source/data/batch_inference.rst) that uses `.. tab-set::` magic.

**Comment by pcmoritz** [comment]
The problem is that the tests are failing with this:
```

[2025-02-08T01:34:27Z] /ray/doc/source/cluster/kubernetes/examples/ml-example.md:14: WARNING: undefined label: 'train-gbdt-guide'
--
  | [2025-02-08T01:34:27Z] /ray/doc/source/cluster/kubernetes/examples/ml-example.md:180: WARNING: undefined label: 'train-gbdt-guide'
  | [2025-02-08T01:34:28Z] /ray/doc/source/cluster/vms/examples/ml-example.md:14: WARNING: undefined label: 'train-gbdt-guide'
  | [2025-02-08T01:34:28Z] /ray/doc/source/cluster/vms/examples/ml-example.md:120: WARNING: undefined label: 'train-gbdt-guide'
  | [2025-02-08T01:34:38Z] /ray/doc/source/ray-overview/use-cases.rst:175: WARNING: unknown document: '/train/examples/xgboost/xgboost_example'
  | [2025-02-08T01:34:50Z] /ray/doc/source/train/more-frameworks.rst:43: WARNING: unknown document: 'distributed-xgboost-lightgbm'


```

(you should be able to see this in the CI run linked from my comment)

**Comment by crypdick** [comment]
@pcmoritz ok, I've fixed 3 of the error categories. 

I'm stuck on the `undefined label` errors from the missing anchors. The [Sphinx cross-referencing docs](https://docs.readthedocs.com/platform/stable/guides/cross-referencing-with-sphinx.html) don't mention ipynb support. However, it is possible to [link to arbitrary webpages](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages), would that be a satisfactory work-around?

Alternatively, I [found an ugly hack](https://github.com/cuthbertLab/music21/blob/master/documentation/docbuild/writers.py) to create anchors by converting the ipynb notebooks, which I don't recommend.

**Comment by pcmoritz** [comment]
Does it work if you replace the labels `train-gbdt-guide` with `/train/examples/xgboost/distributed-xgboost-lightgbm`? Unfortunately using link to arbitrary webpages is not a valid workaround (first, it doesn't check the webpage exists, and more importantly, it doesn't handle versioning of the docs correctly)

```

[2025-02-10T21:53:31Z] /ray/doc/source/cluster/kubernetes/examples/ml-example.md:14: WARNING: undefined label: 'train-gbdt-guide'
--
  | [2025-02-10T21:53:31Z] /ray/doc/source/cluster/kubernetes/examples/ml-example.md:180: WARNING: undefined label: 'train-gbdt-guide'
  | [2025-02-10T21:53:32Z] /ray/doc/source/cluster/vms/examples/ml-example.md:14: WARNING: undefined label: 'train-gbdt-guide'
  | [2025-02-10T21:53:32Z] /ray/doc/source/cluster/vms/examples/ml-example.md:120: WARNING: undefined label: 'train-gbdt-guide'


```

**Comment by crypdick** [comment]
@pcmoritz now getting `WARNING: undefined label: '/train/examples/xgboost/distributed-xgboost-lightgbm'` 😭 I feel like there has to be a way

File: doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb
Comment (crypdick) [reply doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: resolving as per our chat

Comment (crypdick) [reply doc/source/train/examples/xgboost/distributed-xgboost-lightgbm.ipynb:1]: resolving as per our chat

**Review by pcmoritz — APPROVED** [review]
Very nice! The only small wrinkle is that now this is displayed twice in the example gallery, maybe we can deduplicate it going forward (but not a big deal):

Screenshot 2025-02-11 at 4 47 54 PM
